examples/src/main/java/com/datastax/oss/driver/examples/mapper/killrvideo/user/LoginQueryProvider.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.examples.mapper.killrvideo.user;
+
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.api.core.cql.Row;
+import com.datastax.oss.driver.api.mapper.MapperContext;
+import com.datastax.oss.driver.api.mapper.entity.EntityHelper;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Provides the implementation of {@link UserDao#login}.
+ *
+ * <p>Package-private visibility is sufficient, this will be called only from the generated DAO
+ * implementation.
+ */
+class LoginQueryProvider {
+
+  private final CqlSession session;
+  private final EntityHelper<User> userHelper;
+  private final PreparedStatement preparedSelectCredentials;
+  private final PreparedStatement preparedSelectUser;
+
+  LoginQueryProvider(
+      MapperContext context,
+      EntityHelper<User> userHelper,
+      EntityHelper<UserCredentials> credentialsHelper) {
+
+    this.session = context.getSession();
+
+    this.userHelper = userHelper;
+
+    this.preparedSelectCredentials =
+        session.prepare(credentialsHelper.selectByPrimaryKey().asCql());
+    this.preparedSelectUser = session.prepare(userHelper.selectByPrimaryKey().asCql());
+  }
+
+  Optional<User> login(String email, char[] password) {
+    return Optional.ofNullable(session.execute(preparedSelectCredentials.bind(email)).one())
+        .flatMap(
+            credentialsRow -> {
+              String hashedPassword = credentialsRow.getString("password");
+              if (PasswordHashing.matches(password, hashedPassword)) {
+                UUID userid = credentialsRow.getUuid("userid");
+                Row userRow = session.execute(preparedSelectUser.bind(userid)).one();
+                if (userRow == null) {
+                  throw new IllegalStateException(
```

Comment:
This method works in a functional way but we are throwing an exception from the inside of it.
Maybe we could Encapsulate user in a `Try` construct? We could use https://www.vavr.io/vavr-docs/#_try but I guess this is not worth adding an additional dependency

Reply:
> I guess this is not worth adding an additional dependency

Yeah I agree.